src/librustc/ty/layout.rs

```diff
@@ -356,12 +356,14 @@ impl<'tcx> LayoutCx<'tcx, TyCtxt<'tcx>> {
             debug!("univariant offset: {:?} field: {:#?}", offset, field);
             offsets[i as usize] = offset;
 
-            if let Some(mut niche) = field.largest_niche.clone() {
-                let available = niche.available(dl);
-                if available > largest_niche_available {
-                    largest_niche_available = available;
-                    niche.offset += offset;
-                    largest_niche = Some(niche);
+            if !repr.hide_niche() {
+                if let Some(mut niche) = field.largest_niche.clone() {
```

Comment:
I am aware there are other ways to phrase this that avoid injecting rightward drift, like `let largest_niche = if repr.hide_niche() { None } else { field.largest_niche.clone() }; if let Some(mut niche) = largest_niche { ... }`

if there's demand to do something like that, I'll put it in.

Reply:
Imo the problem is the language (a lack of let chains) and not your code here. Although adding comments re. what the code does could help?

Reply:
IMO the current structure is clearer than the alternative and the rightward drift is no big deal.

Reply:
Bikeshed alternatives:
```rust
if let (false, Some(mut niche)) = (repr.hide_niche(), field.largest_niche.clone()) {
```
```rust
if let Some(mut niche) = field.largest_niche.clone().filter(|_| !repr.hide_niche()) {
```

Reply:
Wait, what am I talking about, the body of this is tiny, nevermind.

Reply:
(oh, believe you me, I spent a while trying to convince myself that `if let (false, Some(mut niche)) = (repr.hide_niche(), ...) { ... }` was the way to go. But I just couldn't deal with having the `false` be that "far" from the `hide_niche()` method call. I even considered making an `enum NicheHidden { ... }` to make the bit's meaning more apparent.)